google/cloud/security/scanner/audit/fw_rules_engine.py

```diff
@@ -14,64 +14,354 @@
+                    rule.id, rule_index))
+        self.rules_map[rule.id] = rule
+
+    def add_rule_groups(self, group_defs):
+        """Creates group to rule matching.
+
+        Args:
+          group_defs (dict): A dictionary with a group id and a list of rule ids
+            that will be included by including this group in a policy.
+
+        Raises:
+          DuplicateFirewallGroupError: Raised if the group id already exists.
+          RuleDoesntExistError: Raised if a rule included in the group does not
+            exist.
+          InvalidGroupDefinition: Raised if a group definition is invalid.
+        """
+        for group_def in group_defs:
+            group_id = group_def.get('group_id')
+            if not group_id:
+                raise InvalidGroupDefinition('Group requires a group id')
+            if group_id in self.rule_groups_map:
+                raise DuplicateFirewallGroupError(
+                    'Group id already exists: %s' % group_id)
+            rule_ids = group_def.get('rule_ids')
+            if not rule_ids:
+                raise InvalidGroupDefinition(
+                    'Group "%s" does not have any rules' % group_id)
+            for rule_id in rule_ids:
+                if rule_id not in self.rules_map:
+                    raise RuleDoesntExistError(
+                        'Rule id "%s" does not exist, cannot be in group' %
+                        rule_id)
+            self.rule_groups_map[group_id] = group_def.get(
+                'rule_ids', [])
+
+    def add_org_policy(self, org_def):
+        """Creates org policy and rule mapping.
+
+        Sample org structure:
+
+                org 1234
+               /        \
+              f-1       p-c
+             /  \
+           p-a  p-b
+
+        Rules can be applied at any node above. When a policy is being audited,
+        it the rulebook will start at the lowest level (the project) and will
+        walk up the hierarchy until it reaches the first instance with rules and
+        these are the only rules that are checked.
+
+        Args:
+          org_def (dict): A dictionary of resource ids and enforced rules.
+
+        Raises:
+          RuleDoesntExistError: Raised if a rule included in the group does not
+            exist.
+          GroupDoesntExistError: Raised if a group included in an org policy
+            does not exist.
+          InvalidOrgDefinition: Raised if org policy doesn't have resources.
+        """
+        resources = org_def.get('resources', [])
+        if not resources:
+            raise InvalidOrgDefinition('Org policy does not have any resources')
+        for resource in resources:
+            resource_type = resource_mod.ResourceType.verify(
+                resource.get('type'))
+            ids = resource.get('resource_ids', [])
+            rules = resource.get('rules', {})
+            groups = rules.get('group_ids', [])
+            expanded_rules = []
```

Comment:
This should be a set, not a list. That will require adding a __hash__() method to the Rule class that returns hash(self.id) with the explicit expectation that no two Rules can exist in the same rulebook with the same id (which is checked).

Reply:
Good point, done.